parlai/core/torch_generator_agent.py

```diff
@@ -211,13 +214,12 @@ def forward(self, xs, ys=None, cand_params=None, prev_enc=None, maxlen=None):
             self.longest_label = max(self.longest_label, ys.size(1))
 
         # use cached encoding if available
-        encoder_states = prev_enc if prev_enc is not None else self.encoder(xs)
+        encoder_states = prev_enc if prev_enc is not None else self.encoder(*xs)
 
         if ys is not None:
             # use teacher forcing
             scores, preds = self.decode_forced(encoder_states, ys)
         else:
-            bsz = xs.size(0)
             scores, preds = self.decode_greedy(
                 encoder_states,
                 bsz,
```

Comment:
does this protect against bsz being None?

Reply:
When is bsz None? greedy_decode needs to be explicitly given the batch size. It uses `batch.text_vec.size(0)`.

Reply:
I more meant that `bsz` is an optional arg here, and may be `None` - is this ensured anywhere that `bsz` is in fact not `None` when `ys` is? E.g. I found a snippet [here](https://github.com/facebookresearch/ParlAI/blob/a2535496ec454822b83dea7525ceaa4e856dfb22/parlai/core/torch_generator_agent.py#L744) that calls the model with potentially `ys` as `None` and `bsz` as `None`

Reply:
Honestly I think that class is nearly entirely broken... But it's a good point. Let's move this into #1469.